_posts/2018-06-18-On-Obsessive-Statelessness.markdown

```diff
@@ -0,0 +1,203 @@
+Making it not a secure platform for un-trusted code.
+
+Not deterred, I explored the world of docker hosting as a service - first exploring running my own cluster [on AWS
+ECS][ecs] and then settling [on Hyper][hyper] which offered sandboxed runs that booted in a few seconds. This is
+where my first real dive into obsessive statelessness comes in. The docker container, and the hyper environment
+contains no config by default. There is no Peril information available inside that runtime environment.
+
+The information about a Peril run comes exclusively from Peril, in my head, I call this dependency injection for the
+runtime environment. You can get a sense for what the full JSON looks like in this
+[fixtured file generated by tests](https://github.com/danger/peril/blob/master/source/github/events/handlers/_tests/fixtures/PerilRunnerEventBootStrapExample.json).
+It contains everything from (temporary) GitHub access tokens (only for your installation), to environment variables
+for your run and the webhook JSON. The runtime environment only knows that information for the duration of the
+process then all access tokens expires after 30 seconds of it starting regardless.
+
+## User Sessions are ephemeral
+
+With the runtime security figured out, and reasonably stable, I could start thinking about how people can control
+their installations inside peril. Taking ideas from Ashkan's [post on JWT's][jwts] I explored using JWTs to a few
+database-y related problems.
+
+A JWT is a string made of three components, a header which says how it is signed, a base64 chunk of JSON data and a
+signature verifying the data. You can always read the data inside a JWT, but you need to know the public key used to
+sign the token to verify that it's not been tampered with or created from another source.
+
+This brings us to my first problem when building out multi-account Peril: User accounts. For Peril the root element
+of the domain model is a GitHub Installation. An non-reflective perspective on building a web service would have me
+creating a user modal which can keep track of permission and unique user settings. With [GDPR so freshly
+baked][gdpr] I really didn't see any actual value in keeping this kind of data. Instead I added enough metadata to a
+JWT to replace a user model completely.
+
+#### Here's an example JWT
+
+```sh
+eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE1MjkxOTgwOTcsImlzcyI6WyIxMjMiLCIzMjEiXSwiZGF0YSI6eyJ1c2VyIjp7Im5hbWUiOiJPcnRhIiwiYXZhdGFyX3VybCI6Imh0dHBzOi8vYXZhdGFyczIuZ2l0aHVidXNlcmNvbnRlbnQuY29tL3UvNDkwMzg_dj00In19LCJleHAiOjE1MjkyMDE2OTd9.E4NgIYCLdOkGO76fgQ7ERnYGErX90Bg_JpfmoPSmamU8F-6578FQBpEhu9sWo337uIc379L6P61NTKcpiC09jhhTXdESeiVT2-gW3HK1XhI7fRibu42wMi3R-mdv30u9uN75d-2QqQhxitk9Dr6u1zaPeMEVX6c-J9TGX6kVAPas506zn9vbnWrAEgpZnCC7c7-vgdoHOvTZE7ugjvi16GmMfQjKSFDAh43fpgf_2c9Es6RtB1vaTu1TVuVB-Slvu7NbczJQcGRa1fAPdKp8JWvaXXDYrKIDeSC-YJxZZcSqSAuso6UZ8-DJLcHVquEgZVM3AfSVgX04JORzA_G_wg
```

Comment:
Okay so the cause of the problem is [this line of code](https://github.com/brianmario/yajl-ruby/blob/8830754ac1ca04a0c14f71f6e7c1d240934366fe/ext/yajl/yajl_ext.c#L87) in the JSON library used by the Jekyll plugin that does our Lunr.js search. It looks like something about trying to index this JWT causes the index to exceed the [hardcoded and unconfigurable](https://github.com/brianmario/yajl-ruby/blob/c85c858af64d9269a257899e9b9dd995fa0a1b16/ext/yajl/api/yajl_common.h#L40) nesting maximum for JSON objects. 

The easiest solution I think would be to throw this JWT into a gist and then embed/link to that gist from here. You could also add `exclude_from_search: true` to the frontmatter of this post, but then it wouldn't show up in search. No easy answer – sorry!

Reply:
maybe I can switch it to inline code?

Reply:
it doesn't, interresting

Reply:
Yeah it's a fascinating bug – my guess is that because the JWT is so information-dense in a way that the rest of the blog isn't and it just makes Lunr.js go off the rails.

Reply:
Aye, interesting eh? I switched it out with a gist as you recommended, it's a reasonable trade-off IMO